File: mapping/test/systemtests/parse_logs.php
#!/usr/bin/php
<?php

// These are test failures, e.g. the test had the wrong result
$failure_passes = array('semantic', 'hash');
// These are test errors, e.g. the test crashed or did not run correctly
$error_passes = array('returncode', 'elapsedtime', 'sanitizer');
// These passes are reported in the HTML output
$reported_passes = array('semantic', 'hash', 'sanitizer');
define('ROOT', 'test/systemtests/queries/');


if ($argc != 3)
	die("Needs two parameters: the file for HTML output and the file for JUnit XML output\n");

$html_file = $argv[1];
$xml_file = $argv[2];

$timestamp = time();
$datehuman = date('r', $timestamp);
$dateiso = date('c', $timestamp);


// Output as a human-readable HTML page, including logs.
$html_out = <<<EOS
<html>
	<head><title>Testresults $datehuman</title>
	<script type="text/javascript">
	function toggle(id) {
		var e = document.getElementsByClassName('collapse_'+id);
		for (var i=0;i<e.length;i++)
			e[i].style.display = e[i].style.display == 'none' ? '' : 'none';
	}
	</script>
	<style>
   	thead td {
		font-weight: bold;
	}
	td {
		text-align: center;
		padding: 1px 5px;
		margin: 0;
	}
	td:first-child {
		text-align: left;
	}
    </style>
    </head>
	<body>
		<h1>Testresults $datehuman</h1>
		<table border="1" cellspacing="0">
		<thead><tr><td>Name</td><td style="min-width: 70px;">Time</td>
EOS;

foreach($reported_passes as $pass)
	$html_out .= '<td style="min-width: 100px;">' . $pass . '</td>';
$html_out .= '<td style="min-width: 50px;">Exit code</td></tr></thead>' . "\n";

// Output in JUnit's XML format, documented for example here:
// https://github.com/windyroad/JUnit-Schema/blob/master/JUnit.xsd
$xml_out = new DOMDocument('1.0', 'UTF-8');
$xml_out_testsuites = $xml_out->createElement('testsuites');
$xml_out_testsuites->setAttribute('timestamp', $dateiso);
$xml_out_testsuites->setAttribute('name', 'AllSystemTests');
$xml_out->appendChild($xml_out_testsuites);

$xml_out_testsuite = $xml_out->createElement('testsuite');
$xml_out_testsuite->setAttribute('name', 'Systemtests');
$xml_out_testsuites->appendChild($xml_out_testsuite);

// Find all log files
$logs = array();
$dir = opendir(ROOT);
while ($file = readdir($dir)) {
	$path = ROOT . $file;
	if (!is_file($path))
		continue;
	$pi = pathinfo($path);
	if ($pi['extension'] != 'log')
		continue;
	$logs[] = $file;
}
closedir($dir);

sort($logs);


// Parse all logs and generate output
$tests_passed = 0;
$tests_total = 0;
$tests_totaltime = 0;

$collapse_id = 0;
foreach ($logs as $file) {
	$collapse_id++;

	$log = file_get_contents(ROOT . $file);
	$logname = str_replace('.log', '', $file);

	// We need to manually search the log for indications of successes or failures
	$elapsedtime = -1;
	$returncode = -1;
	$passed = array();

	// By default, all passes are FAIL unless the log says otherwise
	foreach($failure_passes as $name)
		$passed[$name] = false;
	foreach($error_passes as $name)
		$passed[$name] = false;

	// These are considered PASS unless the log says otherwise.
	$passed['sanitizer'] = true;

	// Find the returncode and time
	if (preg_match('/^TESTCASE_ELAPSED_TIME: ([.0-9]+)$/m', $log, $matches)) {
		$elapsedtime = doubleval($matches[1]);
		$passed['elapsedtime'] = true;
	}
	if (preg_match('/^TESTCASE_RETURN_CODE: ([0-9]+)$/m', $log, $matches)) {
		$returncode = intval($matches[1]);
		if ($returncode == 0)
			$passed['returncode'] = true;
	}

	// Find indications of PASS or FAIL in the log
	foreach($reported_passes as $pass) {
		if (strpos($log, "PASSED: $pass\n") !== FALSE)
			$passed[$pass] = true;
		if (strpos($log, "FAILED: $pass\n") !== FALSE)
			$passed[$pass] = false;
	}

	// Watch for output from the sanitizers
	if (strpos($log, 'AddressSanitizer') !== FALSE || strpos($log, 'LeakSanitizer') !== FALSE || strpos($log, ': runtime error: ') !== FALSE)
		$passed['sanitizer'] = false;

	// Determine if the whole test is considered PASS or not
	$all_passed = true;
	foreach($passed as $name => $success) {
		if (!$success)
			$all_passed = false;
	}

	$tests_total++;
	if ($all_passed)
		$tests_passed++;
	$tests_totaltime += max(0, $elapsedtime);

	// Generate HTML output
	$html_out .= '<tr'.($all_passed ? '' : ' onclick="toggle('.$collapse_id.');" style="cursor: pointer;"').'><td>'.$logname.'</td>';

	$html_out .= '<td style="text-align: right;">' . number_format($elapsedtime, 3, '.', ' ') . 's</td>';

	foreach($reported_passes as $pass) {
		if (!isset($passed[$pass]))
			$col = 'blue';
		else if ($passed[$pass])
			$col = 'green';
		else
			$col = 'red';
		$html_out .= '<td style="background-color: ' . $col . '">'.$pass.'</td>';
	}
	$html_out .= '<td style="background-color: ' . ($returncode == 0 ? 'green' : 'red') . '">' . $returncode . '</td>';

	$html_out .= "</tr>\n";
	if (!$all_passed) {
		$html_out .= '<tr class="collapse_'.$collapse_id.'" style="display: none"><td colspan="'.(count($reported_passes)+3).'"><b>Execution log</b><br /><pre>'.htmlspecialchars($log, ENT_NOQUOTES).'</pre></td></tr>'."\n";
	}

	// Generate XML output
	$xml_testcase = $xml_out->createElement('testcase');
	$xml_testcase->setAttribute('name', $logname);
	$xml_testcase->setAttribute('classname', 'systemtests.'.$logname);
	$xml_testcase->setAttribute('status', 'run');
	if ($elapsedtime >= 0)
		$xml_testcase->setAttribute('time', $elapsedtime);

	if (!$all_passed) {
		$failed = array();
		$is_error = false;
		foreach($passed as $name => $success) {
			if (!$success) {
				if (in_array($name, $error_passes))
					$is_error = true;
				$failed[] = $name;
			}
		}
		$failure = $xml_out->createElement($is_error ? 'error' : 'failure', $log);
		$failure->setAttribute('message', 'Failed: ' . implode(', ', $failed));
		$xml_testcase->appendChild($failure);
	}

	$xml_out_testsuite->appendChild($xml_testcase);
}

// Finish HTML output
$html_out .= '</table></body></html>';
file_put_contents($html_file, $html_out);

// Finish XML output
// <testsuites tests="163" failures="0" disabled="0" errors="0" timestamp="2016-03-17T13:30:04" time="0.419" name="AllTests">
// <testsuite name="CSVParser" tests="12" failures="0" disabled="0" errors="0" time="0">
function xml_set_aggregates($node, $passed, $total, $time) {
	$node->setAttribute('tests', $total);
	$node->setAttribute('failures', $total-$passed);
	$node->setAttribute('disabled', 0);
	$node->setAttribute('errors', 0);
	$node->setAttribute('time', number_format($time, 3, '.', ''));
}
xml_set_aggregates($xml_out_testsuite, $tests_total, $tests_passed, $tests_totaltime);
xml_set_aggregates($xml_out_testsuites, $tests_total, $tests_passed, $tests_totaltime);
$xml_out->formatOutput = true;
$xml_out->save($xml_file);

File: mapping/test/systemtests/parse_logs.cpp


/*
 * This program takes a bunch of test-logs and outputs the results in JUnit's XML format.
 *
 * A documentation of the XML format can be found here:
 * https://github.com/windyroad/JUnit-Schema/blob/master/JUnit.xsd
 */

#include <iostream>
#include <fstream>
#include <sstream>
#include <iomanip>

#include <string>
#include <vector>
#include <algorithm> // sort(), max()

#include <ctime>

#include <sys/types.h>
#include <unistd.h>
#include <dirent.h> // opendir(), ..


static const char *LOG_ROOT = "test/systemtests/queries/";


// mapping's core system does not require an XML library, and we do not
// want to introduce another dependency just for a testcase helper binary.
// Instead, we're using the good old XML-via-concatenation approach.
static void encodeXML(std::string &out, const std::string &in) {
	for(size_t pos = 0; pos != in.size(); pos++) {
		switch(in[pos]) {
				case '&':  out.append("&amp;");       break;
				case '\"': out.append("&quot;");      break;
				case '\'': out.append("&apos;");      break;
				case '<':  out.append("&lt;");        break;
				case '>':  out.append("&gt;");        break;
				default:   out.append(&in[pos], 1); break;
		}
	}
}

// We are aware of c++11's std::regex, but it's not fully supported in gcc 4.8, which we're using.
// again, we do not want to bump our compiler requirements for a a testcase helper binary.
static bool str_starts_with(const std::string &subject, const std::string &prefix, std::string &suffix) {
	if (subject.substr(0, prefix.length()) == prefix) {
		suffix = subject.substr(prefix.length());
		return true;
	}
	suffix = "";
	return false;
}

static std::string getIsoDate() {
	auto t = std::time(nullptr);
	if (t < 0)
		return "";
	std::tm *tm = std::gmtime(&t);
	if (tm == nullptr)
		return "";

	int year = 1900 + tm->tm_year;
	int month = tm->tm_mon + 1;
	int day = tm->tm_mday;
	int hour = tm->tm_hour;
	int minute = tm->tm_min;
	int second = tm->tm_sec;

	std::ostringstream result;
	result.fill('0');

	result << std::setw(4) << year << "-" << std::setw(2) << month << "-" <<std::setw(2) << day << "T"
	       << std::setw(2) << hour << ":" <<std::setw(2) << minute << ":" <<std::setw(2) << second << "Z";

	return result.str();
}


int main() {
	// Find all log files
	DIR *dir = opendir(LOG_ROOT);
	if (!dir) {
		perror("opendir failed");
		exit(5);
	}

	std::vector<std::string> lognames;
	while (auto dirent = readdir(dir)) {
		std::string name(dirent->d_name);
		if (name.length() < 4 || name.substr(name.length() - 4, 4) != ".log")
			continue;

		lognames.push_back(name.substr(0, name.length()-4));
	}
	closedir(dir);

	std::sort(lognames.begin(), lognames.end());

	int tests_total = 0;
	int tests_errors = 0;
	int tests_failures = 0;
	double tests_totaltime = 0;
	std::string xml;

	for (auto &logname : lognames) {
		// Load and parse the logfile
		std::ifstream stream(std::string(LOG_ROOT) + logname + ".log", std::ios::in | std::ios::binary);
		if (!stream)
			throw std::runtime_error("Could not open logfile");

		std::string full_log;
		bool passed_semantic = false;
		bool passed_hash = false;
		bool passed_sanitizer = true;
		double elapsedtime = -1;
		int returncode = -1;

		// iterate over all lines in the log
		std::string line;
		while (std::getline(stream,  line)) {
			std::string suffix;
			if (str_starts_with(line, "TESTCASE_ELAPSED_TIME: ", suffix)) {
				elapsedtime = std::stod(suffix);
			}
			if (str_starts_with(line, "TESTCASE_RETURN_CODE: ", suffix)) {
				returncode = std::stoi(suffix);
			}
			if (str_starts_with(line, "PASSED: semantic", suffix)) {
				passed_semantic = true;
			}
			if (str_starts_with(line, "PASSED: hash", suffix)) {
				passed_hash = true;
			}
			if (line.find("AddressSanitizer") != std::string::npos || line.find("LeakSanitizer") != std::string::npos || line.find(": runtime error: ") != std::string::npos) {
				passed_sanitizer = false;
			}

			full_log += line + "\n";
		}
		stream.close();

		// find total test status
		// These are test failures, e.g. the test had the wrong result
		bool has_failure = false;
		if (!passed_semantic || !passed_hash)
			has_failure = true;

		// These are test errors, e.g. the test crashed or did not run correctly
		bool has_error = false;
		if (returncode != 0 || !(elapsedtime > 0) || !passed_sanitizer)
			has_error = true;

		// accumulate global stats
		tests_total++;
		if (has_error)
			tests_errors++;
		else if (has_failure)
			tests_failures++;
		tests_totaltime += std::max(0.0, elapsedtime);

		// append to XML output
		xml += "<testcase name=\"";
		encodeXML(xml, logname);
		xml += "\" classname=\"systemtests.";
		encodeXML(xml, logname);
		xml += "\" status=\"run\"";
		if (elapsedtime > 0)
			xml += " time=\"" + std::to_string(elapsedtime) + "\"";
		xml += ">";
		if (has_failure || has_error) {
				xml += (has_error ? "<error" : "<failure");
				xml += " message=\"Failed:";
				if (returncode != 0)
					xml += " returncode";
				if (elapsedtime <= 0)
					xml += " elapsedtime";
				if (!passed_sanitizer)
					xml += " sanitizer";
				if (!passed_semantic)
					xml += " semantic";
				if (!passed_hash)
					xml += " hash";
				xml += "\">";
				encodeXML(xml, full_log);
				xml += (has_error ? "</error>" : "</failure>");
		}

		xml += "</testcase>\n";
	}

	// Finalize XML output
	std::string xml_header;
	xml_header = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n";
	xml_header += "<testsuites timestamp=\"";
	encodeXML(xml_header, getIsoDate());
	xml_header += "\" name=\"AllSystemtests\" tests=\"" + std::to_string(tests_total) + "\" failures=\"" + std::to_string(tests_failures) + "\" disabled=\"0\" errors=\"" + std::to_string(tests_errors) + "\" time=\"" + std::to_string(tests_totaltime) + "\">\n";
	xml_header += "<testsuite name=\"Systemtests\" tests=\"" + std::to_string(tests_total) + "\" failures=\"" + std::to_string(tests_failures) + "\" disabled=\"0\" errors=\"" + std::to_string(tests_errors) + "\" time=\"" + std::to_string(tests_totaltime) + "\">\n";

	std::string xml_footer;
	xml_footer = "</testsuite></testsuites>\n";

	std::cout << xml_header << xml << xml_footer;
}

